spec/lib/granite/context_proxy_spec.rb

```diff
@@ -0,0 +1,92 @@
+# spec: unit
+
+require 'granite/context_proxy'
+
+RSpec.describe Granite::ContextProxy do
+  subject { klass.new }
+  let(:klass) do
+    Class.new do
+      include Granite::ContextProxy
+
+      def self.hash
+        'hash'
+      end
+    end
+  end
+  let(:performer) { performers.first }
+  let(:performers) { 2.times.map { instance_double(User) } }
+  let(:context) { {performer: performer} }
+  let(:proxy) { instance_double(Granite::ContextProxy::Proxy) }
+
+  describe '.using' do
+    before do
+      allow(Granite::ContextProxy::Proxy).to receive(:new).with(klass, have_attributes(**context)) { proxy }
+    end
+
+    specify do
+      expect(klass.using(context)).to eq proxy
+    end
+  end
+
+  describe '.as' do
+    before do
+      allow(Granite::ContextProxy::Proxy).to receive(:new).with(klass, have_attributes(**context)) { proxy }
+    end
+
+    specify do
+      expect(klass.as(performer)).to eq proxy
+    end
+  end
+
+  describe '.with_context' do
+    specify do
+      expect { |b| klass.with_context(context, &b) }.to yield_with_no_args
+    end
+
+    it 'sets proxy_content inside block' do
+      klass.with_context(context) do
+        expect(klass.proxy_context).to eq context
+      end
+      expect(klass.proxy_context).to be_nil
+    end
+
+    it 'correctly works with nested contexts' do
+      klass.with_context(context) do
+        expect(klass.proxy_context).to eq context
+        klass.with_context(performer: performers.second) do
+          expect(klass.proxy_context).to eq(performer: performers.second)
+        end
+        expect(klass.proxy_context).to eq context
+      end
+    end
+
+    it 'keeps context in thread safe way' do
+      thread_performers = []
+      threads = 2.times.map do |i|
+        Thread.new do
+          klass.with_context(performer: performers[i]) do
+            Thread.stop
+            thread_performers << klass.proxy_context[:performer]
+          end
+        end
+      end
+      sleep 0.01 until threads.all?(&:stop?)
+      threads.each { |thread| thread.run.join }
```

Comment:
because you `join` the thread, you are waiting for the first one to finish before `run`ning the next one. Am I right?

```suggestion
      threads.each { |thread| thread.run }
      threads.each { |thread| thread.join }
```

Reply:
No I need for them to run in this order:
1. Thread 1 starts
2. Thread 1 updates `proxy_context`
3. Thread 1 stops
4. Thread 2 starts
5. Thread 2 updates `proxy_context`
6. Thread 2 stops
7. Main thread no longer sleeping since both threads `threads.all?(&:stop?)` is true
8. Thread 1 is resumed and adds `proxy_context` to `thread_performers`
9. Thread 1 finishes
10. Thread 2 is resumed and adds `proxy_context` to `thread_performers`
11. Thread 2 finishes

If we don't use thread safe code, then 2nd thread updates 1st thread's `proxy_context` in step 5 and that means 1st thread adds incorrect results in step 8.